### PR TITLE
Share VS Code settings/extensions nicely

### PR DESCRIPTION
**Pull request by samestep**
This is a second attempt at #51214. It should achieve the same goals with (as far as I can tell) no disadvantages, but the advantages are a bit less pronounced than in the more dictatorial approach that #51214 took:

- Unfortunately, I was unable to figure out how to include [the `mypy` configuration given in the docstring of `tools.mypy_wrapper.main`](https://github.com/pytorch/pytorch/blob/7115a4b8707115b81417481fb6240617ae934806/tools/mypy_wrapper.py#L81-L89), because as @walterddr pointed out, `"${env:HOME}/miniconda3/envs/pytorch/bin/python"` is not guaranteed to be correct on everyone's machine:
  ```json
  {
    "python.linting.enabled": true,
    "python.linting.mypyEnabled": true,
    "python.linting.mypyPath": "${env:HOME}/miniconda3/envs/pytorch/bin/python",
    "python.linting.mypyArgs": [
      "${workspaceFolder}/tools/mypy_wrapper.py"
    ]
  }
  ```

  Importantly, this does not work:
  ```json
  "python.linting.mypyPath": "${workspaceFolder}/tools/mypy_wrapper.py"
  ```
  This is because VS Code does not run the given `mypy` command inside of the user's specified virtual environment, so for instance, on my system, setting the `mypy` command to directly call `tools/mypy_wrapper.py` results in using `mypy 0.782` instead of the correct `mypy 0.812`.

  Sadly, [this](https://code.visualstudio.com/docs/editor/variables-reference#_configuration-variables) does not work either, although I'm not sure why:
  ```json
  {
    "python.linting.mypyPath": "${config:python.pythonPath}",
    "python.linting.mypyArgs": [
      "${workspaceFolder}/tools/mypy_wrapper.py"
    ]
  }
  ```

- As a result, `git clean -fdx; tools/vscode_settings.py` still results in some loss of useful configuration.

One other thing to note: as `.vscode/settings_recommended.json` shows, there are some configuration sections that only take effect within the context of a `"[language]"`, so currently, if a dev already has one of those settings, it would be entirely overwritten by `tools/vscode_settings.py` rather than a graceful merge. This could probably be fixed by using a deep merge instead of the current shallow merge strategy.

**Test plan:**

If you want, you can typecheck the small script added by this PR (no output is expected):
```sh
tools/mypy_wrapper.py $PWD/tools/vscode_settings.py
```
You can also try running it to update your own VS Code workspace settings:
```sh
tools/vscode_settings.py
```
This should have minimal impact on your existing `tools/settings.json` file other than enabling the few explicitly recommended settings (e.g. it should not reorder or remove any of your existing settings).

### CODE REVIEW

**Comment by samestep** [comment]
ah wait, I have an idea for how to solve the `mypy` thing

**Comment by samestep** [comment]
> ah wait, I have an idea for how to solve the `mypy` thing

it only builds on this and might have some drawbacks, though, so I'll do that in a followup PR if this one is merged

File: .gitignore [hunk 1]
```diff
@@ -257,7 +257,7 @@ cmake-build-debug
 # Below files are not deleted by "setup.py clean".
 
 # Visual Studio Code files
-.vscode
+/.vscode/settings.json
```
Comment (walterddr) [inline .gitignore:260]: `launch.json` and `c_cpp_properties.json` are now not ignored.

Comment (samestep) [reply .gitignore:260]: I don't have those on my system; could you post their contents here (perhaps in a paste)? I don't think it's necessarily a bad thing to have those in the repo (thinking mostly of `launch.json`, I dunno what `c_cpp_properties.json` is), but I'm very curious to see their contents on your machine

Comment (samestep) [reply .gitignore:260]: If it turns out that there are too many files that we want to ignore and we don't want to list them all, we could instead do something like this:
```gitignore
/.vscode/*
!/.vscode/extensions.json
!/.vscode/settings_recommended.json
```

Comment (samestep) [reply .gitignore:260]: or if we do decide to ignore `launch.json` and `c_cpp_properties.json` but there aren't any others that we need to ignore, we could just add these two lines:
```gitignore
/.vscode/c_cpp_properties.json
/.vscode/launch.json
```

**Review by walterddr — APPROVED** [review]

File: .gitignore [hunk 2]
```diff
@@ -257,8 +257,10 @@ cmake-build-debug
 # Below files are not deleted by "setup.py clean".
 
 # Visual Studio Code files
-.vscode
 .vs
+/.vscode/*
+!/.vscode/extensions.json
```
Comment (driazati) [inline .gitignore:262]: `git clean -xfd` would still blow away vscode settings right? Maybe we should encourage people to use `python setup.py clean` instead

Comment (samestep) [reply .gitignore:262]: it already does that even before this PR but yeah, that sounds like a good idea to me; in that case could you open a PR to replace `git clean -xfd` with `python setup.py clean` in `CONTRIBUTING.md`?

**Review by malfet — APPROVED** [review]

**Pull request closed without merging**